### PR TITLE
Kodi VOD & VLC Transcoding

### PR DESCRIPTION
**Pull request by wmariz**
VOD works in two different ways, first, loading video file directly from Kodi http vfs (called direct mode), second, VLC can load and convert the video in 3 different profiles (called transcode mode).

On both modes, it will play the video on the browser using a modified version of MediaElements.js with added fallback to VLC plugin.

MediaElements.js can play videos using HTML5, Flash or Silverlight.

VLC transcode can work in two ways: first playing on VLC with output redirected to HTTP or RTSP stream, this mode give us access to video information like build-in tracks but can do only one transcode at time, second with VLM interface that allow us to have multiple simultaneous streams, but have limited information about the video file (I choose the VLM alternative for this project).

VLC plugin, on direct mode, can switch subtitles and audio tracks built-in video files, unfortunately, due to VLM limitation it's not possible when transcoding.

Since MediaElements.js cannot switch tracks on HTML5, Flash and Silverlight and VLC can do it only in direct mode, seems better just remove track support from it, giving more consistency between all the play methods.

IE does not accept HTTP basic authentication, so it can't load Kodi videos (in direct mode), even with a HTTP redirect, that does not work with VLC plugin anyway. The solution is to use a proxy server, or build a proxy within python, but IE is also giving me problems with VLC plugin, since (at least on my computer) it opens the video in a new "Active 3D Output" window. I'm really tempted to leave IE behind.

Also, there are some things I still need to do.

Full screeen support is not working yet. MediaElements.js can do it, but I need to add support on VLC fallback.

VLC needs to work in windowless mode to be able to have MediaElements.js interface on top of it, but in this mode VLC cannot resize without stopping an restarting play. I need to make it continue from the point were it have stopped.

Also when switching between direct mode and transcode profiles, it need to continue playing from the same point.

When transcoding, seek must be implemented as a command sent to VLC VLM.

I need to make more tests with transcode settings and muxers to remove muxers that doesn't work and create default profile for normal users.

If somebody wants to join this project and give me a help, I'll be glad.

### CODE REVIEW

**Comment by Hellowlol** [comment]
I have no objections dropping ie support for certain features. I'm tired of the entire thing. I have used countless of hours adding support for trivial stuff just for ie. (This might be cuz I'm a newb but still...)

As this a rather large pr I will take some time before have tested everything. Can you make a step by step just on how to get this working so we can post this to the forum thread to get more testers?

**Comment by wmariz** [comment]
Hellowlol,

You must have VLC installed for testing the player fallback and to do transcoding.

To transcode you must run VLC in background with the command "vlc --intf http --http-port=<interface-port> --http-password=<password> --rtsp-port=<transcode-port>

It's all you need to setup.

**Comment by Hellowlol** [comment]
Can you add the charger and pycaption to the libs folder?

**Comment by wmariz** [comment]
Sorry, I don't know why Chardet was not included on the commits.

Chardet is needed because Kodi HTTP does not implement a header with the encoding of the files.

Pycaption has a lot of dependencies and should be installed by pip (so I added it to requirements.txt)

Pycaption is really optional, since srt and webvtt works whitout any conversion.

**Comment by Hellowlol** [comment]
Thanks. Feel free to make a post in the support thread on kodi forums asking for testers. I'll merge the current pr in a temp branch to make it easier for them.

File: modules/kodi.py [hunk 1]
```diff
@@ -55,7 +72,153 @@ def __init__(self):
...
+                 'value': 'Low',
+                 'name': 'kodi_vlc_transcode3_name'},
+                {'type': 'text',
+                 'label': 'Transcode Settings',
+                 'value': 'vcodec=h264,vb=800,width=640,height=480,acodec=mp3,ab=128,channels=2,samplerate=44100',
+                 'name': 'kodi_vlc_transcode3_settings'},
+                {'type': 'select',
+                 'label': 'Muxer',
+                 'name': 'kodi_vlc_transcode3_muxer',
+                 'desc': '<BR><a href="' + htpc.settings.get('app_webdir') + 'kodi/managevlc">Manage VLC transcode instances</a>',
```
Comment (Hellowlol) [inline modules/kodi.py]: This line has to be fixed, link dosnt work with webdir set.

**Comment by Hellowlol** [comment]
So how can i access any of the files? I can only find manage vlc. How can i start direct play?

The labels in the settings page could use some work, i have to inspect the source to see what the "port" is to.

File: interfaces/default/html/kodi.html [hunk 2]
```diff
@@ -163,4 +163,19 @@ <h1 class="page-header page-title">
         </div>
         <div class="spinner"></div>
     </div>
-</div>
\ No newline at end of file
+</div>
+<script>
+  % if settings.get('kodi_vod_enabled'): 
+    kodi_vod_enabled = true;
```
Comment (Hellowlol) [inline interfaces/default/html/kodi.html]: I needed to add var to the variables to get this to work in chrome.

File: interfaces/default/js/kodi.js [hunk 3]
```diff
@@ -217,6 +217,13 @@ function loadMovies(options) {
 
                     movieAnchor.append($('<h6>').addClass('title').html(shortenText(movie.title, 12)));
 
+                    if (kodi_vod_enabled){
+                        movieItem.append($('<h6>').addClass('title').append($('<a>').attr('href', '#').html("Play in Browser").click(function(e) {
```
Comment (Hellowlol) [inline interfaces/default/js/kodi.js]: I would like to remove this and rather have a button in the modal. What do you think about that?

**Comment by wmariz** [comment]
I also prefer a button in the modal. I will change that and the other suggestions you made

After your first tests and the temporary branch I'll post it on Kodi forums.

**Comment by Hellowlol** [comment]
Thanks for the latest  pr, i have only tested direct play so far, and i couldnt get it to work. (it was a mkv and i assume its not supported and needs to be transcoded.) Ill download some mp4 to test and test transcoding later today.

**Comment by wmariz** [comment]
Hellowlol, 

Virtually any kind of videos are supported in direct mode, but VLC web plug-in need to be installed, since without it the support is limited by the formats MediaElement.js can handle (http://mediaelementjs.com/#devices).

You can test if the plugin is installed and working here http://blog.revolunet.com/VLCcontrols/src/simple.html

**Comment by Hellowlol** [comment]
So just to be clear the web plugin is a plugin to vlc or a browser extension? (i done have access to a pc now)

**Comment by wmariz** [comment]
It's a plugin, like flash and silverlight. It's installed along with VLC.

If you are using a 32bit browser, even with 64bit system, you should install VLC 32bit version.

I'm using the last nightly since 2.2.0 stable have a problem with h264 transcoding.

http://nightlies.videolan.org/build/win32/last/

**Comment by Hellowlol** [comment]
I cant get direct play to work inside kodiplayer however if i paste the url manually to a new tab i can play it.

**Comment by wmariz** [comment]
Ok,

First check if your browser is not asking permission to run VLC, maybe on the popup window you can't see the warning.

Second, check if HTPC Manager can communicate with VLC, in "Manage VLC" it shows a error or just says there are no transcode instances running (as expected)?

Also check if you really put 8001 on HTTP transcode port, since it only have a placeholder, without default value.

Beside that, you can try to put VLC interface ip and port on browser, you should see the web interface (leave login blank, use just the password).

If you try to put the ip and transcode port on the browser you should also see a VLC 404 error.

Htpc Manager will communicate with VLC VLM interface on "HTTP interface IP/Host" at "Interface port" (command line switch --http-port) and the transcoding will be available on "HTTP transcode port" (command line switch --rtsp-port). Password is also required.

Third, this version (2.2.1) have a problem with h264 transcoding (my mistake when I said 2.2.0). Try the nightly  build of VLC or the medium profile.

The high profile uses Webm that apparently is not compatible with VLC http streaming (as I said, I need to do more tests).

The medium profile uses Ogg that can play on MediaElement.js HTML5 player.

The low profile user Flv that can play on Flash fallback, but it also uses h264 encoding that is broken in 2.2.1 stable.

See that 2.2.2 will be released in a few weeks.

**Comment by Hellowlol** [comment]
Sorry i got direct playing working before you posted using the vlc plugin (choppy picture), transcode on medium works aswell (grainy images). Its cool!
the tr should be fixed. you can add the mofa or ofa class to the parent div.
image

**Comment by wmariz** [comment]
What was the problem with vlc plug-in? How do you solve it? I think it's important to know, because users will probably have the same problem.

Also, do you think this will give us much headache? It's to much for normal users?

What are you thoughts? Can something make it easier?

**Comment by Hellowlol** [comment]
Im not really sure, it suddenly worked... only thing i can remember that was different was that i closed vlc then, ran the command (no gui visable other then in taskmgr). 

While it could be a pita to setup and getting it right i think its cool awesome if the playback gets stable.

Well if they want something easy they can just use plex or mb3. 

What we need to do is add a checklist regarding vlc to the faq like you give to me. I pretty sure there will be some questions and i dont want to repeat the same thing over and over.

**Comment by wmariz** [comment]
I think this discussion will be a great faq, because of that, I chose to wait your tests, before posting on forums.

I had trouble with the nightly and avast, it puts VLC in the sandbox on the first runs

**Comment by Hellowlol** [comment]
I agree, but i dont think a normal user will think to check for comments on a "old unmerged pr" so i added the important parts it to https://github.com/Hellowlol/HTPC-Manager/wiki/Streaming
It need to cleaned up, but we can do that later.

**Pull request closed without merging**